### PR TITLE
Replace node objects with methods

### PR DESCRIPTION
**Pull request by joeldrapper**
Early on, I created various types of node objects: components, standard elements, void elements, text nodes, etc. This made it easy to read the code and build a picture in your head of what was going on. But all these nodes are object allocations — big ones, too. And they're not necessary.

This PR removes all node types apart from component and instead uses methods to concat on the buffer string. The component node is updated to be faster too be removing the step where it stores its own content buffer string before pushing it on the main buffer.

The result of this PR is a 44% performance improvement on our main benchmark.

Before:
```
ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [arm64-darwin21]
Warming up --------------------------------------
                Page    25.000  i/100ms
Calculating -------------------------------------
                Page    249.084  (± 0.4%) i/s -      1.250k in   5.018461s
```

After:
```
ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [arm64-darwin21]
Warming up --------------------------------------
                Page    36.000  i/100ms
Calculating -------------------------------------
                Page    359.743  (± 0.3%) i/s -      1.800k in   5.003611s
```

### CODE REVIEW

File: lib/phlex/block.rb [hunk 1]
```diff
@@ -10,7 +10,7 @@ def initialize(context, &block)
     end
 
     def call(target, *args, **kwargs)
-      @context.render_block(target, *args, **kwargs, &@block)
+      @context._render_block(target, *args, **kwargs, &@block)
```
Comment (joeldrapper) [inline lib/phlex/block.rb:13]: Renamed this as it's a private method.

File: lib/phlex/component.rb [hunk 2]
```diff
@@ -1,9 +1,9 @@
 # frozen_string_literal: true
 
-require "digest"
```
Comment (joeldrapper) [inline lib/phlex/component.rb:3]: This was left over from the caching checksum stuff we removed #45.

File: lib/phlex/context.rb [hunk 3]
```diff
@@ -2,16 +2,20 @@
 
 module Phlex
   module Context
+    def content(&)
```
Comment (joeldrapper) [inline lib/phlex/context.rb:5]: This is a DSL method so felt more at home here on Context.

File: lib/phlex/component.rb [hunk 4]
```diff
@@ -62,25 +60,9 @@ def call(buffer = +"")
...
       @_rendered = true
 
+      @_target = buffer
       template(&@_content)
-      super
-    end
-
-    def target
-      @_target || self
-    end
```
Comment (joeldrapper) [inline lib/phlex/component.rb:71]: We always reference `@_target` directly to save a method call.

File: lib/phlex/context.rb [hunk 5]
```diff
@@ -2,16 +2,20 @@
...
+    end
+
     def text(content)
-      Text.new(content).call(target.children)
+      @_target << CGI.escape_html(content)
     end
 
     def whitespace
-      _raw(" ")
+      @_target << " "
```
Comment (joeldrapper) [inline lib/phlex/context.rb:14]: Skip the `_raw` method call.

File: lib/phlex/context.rb [hunk 6]
```diff
@@ -24,12 +28,21 @@ def component(component, *args, **kwargs, &block)
...
+        component.new(**kwargs) { text(args.first) }.call(@_target)
       else
-        component.new(*args, **kwargs, &block).call(target.children)
+        component.new(*args, **kwargs, &block).call(@_target)
       end
     end
 
+    def _render_block(new_target, ...)
+      old_target = @_target
+      @_target = new_target
```
Comment (alexandreruban) [inline lib/phlex/context.rb:39]: Do we still need those? Now `@_target` is always going to be the same buffer right?

**Review by alexandreruban — APPROVED** [review]
This looks really clear to me, and impressive performance gain!! 🤯

File: lib/phlex/context.rb (hunk 6 above)
Comment (joeldrapper) [reply lib/phlex/context.rb:39]: You’re right, we don't need it.

**Pull request merged**